File: server/models/shape.go
package models

type Shape struct {
	Name        string `json:"name,omitempty"`
	Description string `json:"desc,omitempty"`

	X int64 `json:"x,omitempty"`
	Y int64 `json:"y,omitempty"`

	Width  int64 `json:"width,omitempty"`
	Height int64 `json:"height,omitempty"`

	Rotation int     `json:"rotation,omitempty"`
	Points   []Point `json:"points,omitempty"`

	Radius  int64 `json:"radius,omitempty"`
	RadiusX int64 `json:"radiusX,omitempty"`
	RadiusY int64 `json:"radiusY,omitempty"`

	StartAngle int64 `json:"startAngle,omitempty"`
	EndAngle   int64 `json:"endAngle,omitempty"`

	BackgroundColor string `json:"backgroundColor,omitempty"`

	BorderColor string `json:"borderColor,omitempty"`
	BorderWidth int64  `json:"borderWidth,omitempty"`

	Src       string `json:"src,omitempty"`
	Mirror    bool   `json:"mirror,omitempty"`
	ProjectId int64  `json:"projectId,omitempty"`
}

File: client/src/models/canvas/shapes/base/shape.ts
import { ShapeTypes } from "./constants";
import { DrawOptions2D, Renderable } from "./renderable";

/**
 * Base object represented in 2D space
 */
export default abstract class Shape implements Renderable {
  name: string;
  desc: ShapeTypes;
  x: number;
  y: number;
  width: number;
  height: number;
  backgroundColor: string;
  borderColor: string;
  borderWidth: number;
  rotation: number;
  mirror: boolean;

  constructor(
    name: string,
    desc: ShapeTypes,
    x: number,
    y: number,
    width: number,
    height: number,
    backgroundColor: string,
    borderColor: string,
    borderWidth: number,
    rotation: number,
    mirror: boolean = false
  ) {
    this.desc = desc;
    this.name = name ?? this.desc;
    this.x = x ?? 0;
    this.y = y ?? 0;
    this.width = width ?? 0;
    this.height = height ?? 0;
    this.backgroundColor = backgroundColor;
    this.borderColor = borderColor;
    this.borderWidth = borderWidth ?? 0;
    this.rotation = rotation ?? 0;
    this.mirror = mirror ?? false;
  }

  abstract draw(context: CanvasRenderingContext2D, options: DrawOptions2D): void;
  abstract drawResized(context: CanvasRenderingContext2D, resizeSize: number, options: DrawOptions2D): void;
}
